src/tools/wasm-api-gen.cpp

```diff
@@ -0,0 +1,233 @@
+
+#define DELEGATE_START(id)                                                     \
+  std::cout << "BINARYEN_API BinaryenExpressionRef\n"                          \
+            << "Binaryen" << #id << "(BinaryenModuleRef module";
+
+#define DELEGATE_FIELD_CHILD(id, name)                                         \
+  std::cout << ", BinaryenExpressionRef " << #name
+
+#define DELEGATE_FIELD_INT(id, name) std::cout << ", uint32_t " << #name
+
+#define DELEGATE_FIELD_INT_ARRAY(id, name)                                     \
+  std::cout << ", const uint8_t mask[16] " << #name
+
+#define DELEGATE_FIELD_LITERAL(id, name)                                       \
+  std::cout << ", struct BinaryenLiteral " << #name
+
+#define DELEGATE_FIELD_NAME(id, name) std::cout << ", const char* " << #name
+
+#define DELEGATE_FIELD_NAME_VECTOR(id, name)                                   \
+  std::cout << ", const char** " << #name << ", BinaryenIndex num_" << #name;
+
+#define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name)                                \
+  std::cout << ", const char* " << #name
+
+#define DELEGATE_FIELD_SCOPE_NAME_USE(id, name)                                \
+  std::cout << ", const char* " << #name
+
+#define DELEGATE_FIELD_SCOPE_NAME_USE_VECTOR(id, name)                         \
+  std::cout << ", const char** " << #name << ", BinaryenIndex num_" << #name;
+
+#define DELEGATE_FIELD_SIGNATURE(id, name)                                     \
+  std::cout << ", BinaryenType " << #name << "_params, BinaryenType "          \
+            << #name "_results";
+
+#define DELEGATE_FIELD_TYPE(id, name) std::cout << ", BinaryenType " << #name
+
+#define DELEGATE_FIELD_ADDRESS(id, name) std::cout << ", uint32_t " << #name
+
+#define DELEGATE_FIELD_CHILD_VECTOR(id, name)                                  \
+  std::cout << ", BinaryenExpressionRef* " << #name << ", BinaryenIndex num_"  \
+            << #name;
+
+#include "wasm-delegations-fields.h"
+
+  if (hasExtraType<T>()) {
+    std::cout << ", BinaryenType type";
+  }
+  if (hasExtraHeapType<T>()) {
+    std::cout << ", BinaryenType heapType";
+  }
+
+  std::cout << ')';
+
+  if (!impl) {
+    std::cout << ";\n";
+  }
+}
+
+template<typename T> void autogenOneCAPIImpl() {
+  autogenOneCAPIDecl<T>(true /* impl */);
+  std::cout << " {\n";
+
+  MixedArena allocator;
+  T curr(allocator);
+
+  std::string className;
+  std::vector<std::string> params;
+
+#define DELEGATE_ID curr._id
+
+#define DELEGATE_START(id) className = #id;
+
+#define DELEGATE_FIELD_CHILD(id, name) params.push_back(#name);
+
+#define DELEGATE_FIELD_INT(id, name)                                           \
+  if (hasOp<T>()) {                                                            \
+    params.push_back(#id "Op(" #name ")");                                     \
+  } else {                                                                     \
+    params.push_back(#name);                                                   \
+  }
```

Comment:
Some SIMD instructions have both op and other int operands, so this won't scale. Perhaps Op operands should get their own delegate to differentiate them?

Reply:
Good point. Yeah, we'll need to be more careful as we roll this out to replace the handwritten ones, which I think we should do eventually.

A new delegate is one option. Another might be to use the order, as the `op` is always first, so we could do that if-else only on the first item in the list.